mesonbuild/mintro.py

```diff
@@ -125,7 +130,128 @@ def list_target_files(target_name, coredata, builddata):
         out.append(i)
     print(json.dumps(out))
 
-def list_buildoptions(coredata, builddata):
+class BuildoptionsOptionHelper:
+    # mimic an argparse namespace
+    def __init__(self, cross_file):
+        self.cross_file = cross_file
+        self.native_file = None
+        self.cmd_line_options = {}
+
+class BuildoptionsInterperter(astinterpreter.AstInterpreter):
+    # Interpreter to detect the options without a build directory
+    # Most of the code is stolen from interperter.Interpreter
+    def __init__(self, source_root, subdir, backend, cross_file=None):
+        super().__init__(source_root, subdir)
+
+        options = BuildoptionsOptionHelper(cross_file)
+        self.environment = environment.Environment(source_root, None, options)
+        self.subproject = ''
+        self.coredata = self.environment.get_coredata()
+        self.option_file = os.path.join(self.source_root, self.subdir, 'meson_options.txt')
+        self.default_project_options = {} # Used for subproject handling. Maybe implement this
+        self.backend = backends.get_backend_from_name(backend, None)
+        self.default_options = {'backend': self.backend.name}
+
+        self.funcs.update({
+            'project': self.func_project,
+            'add_languages': self.func_add_languages
+        })
+
+    def detect_compilers(self, lang, need_cross_compiler):
+        comp, cross_comp = self.environment.detect_compilers(lang, need_cross_compiler)
+        if comp is None:
+            raise RuntimeError('Tried to use unknown language "%s".' % lang)
+
+        self.coredata.compilers[lang] = comp
+        # Native compiler always exist so always add its options.
+        new_options = comp.get_options()
+        if cross_comp is not None:
+            self.coredata.cross_compilers[lang] = cross_comp
+            new_options.update(cross_comp.get_options())
+
+        optprefix = lang + '_'
+        for k, o in new_options.items():
+            if not k.startswith(optprefix):
+                raise RuntimeError('Internal error, %s has incorrect prefix.' % k)
+            self.coredata.compiler_options.setdefault(k, o)
+
+        return comp, cross_comp
+
+    def add_languages(self, args, required):
+        success = True
+        need_cross_compiler = self.environment.is_cross_build() and self.environment.cross_info.need_cross_compiler()
+        for lang in sorted(args, key=compilers.sort_clink):
+            lang = lang.lower()
+            if lang in self.coredata.compilers:
+                comp = self.coredata.compilers[lang]
+            else:
+                (comp, _) = self.detect_compilers(lang, need_cross_compiler)
+            self.add_base_options(comp)
+        return success
+
+    def func_project(self, node, args, kwargs):
+        if len(args) < 1:
+            raise InvalidArguments('Not enough arguments to project(). Needs at least the project name.')
+
+        # Resolve mparser.ArrayNode if needed
+        good_args = []
```

Comment:
`good_args` sounds a bit strange. Maybe `flattend_args` or something like that?

Maybe even append directly to proj_langs. It seems using an array as project name would confuse this currently. That could be avoided by looping over args[1:] and placing the results directly in proj_langs.